intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py

```diff
@@ -115,110 +140,99 @@ def KB_construct(self, input):
+            documents = []
+            for data, meta in data_collection:
+                metadata = {"source": meta}
+                if len(data) < 5:
+                    continue
+                new_doc = SDocument(content=data, meta=metadata)
+                documents.append(new_doc)
+            assert documents != [], "The given file/files cannot be loaded."
+            document_store.write_documents(documents)
+            print("The local knowledge base has been successfully built!")
+            return document_store
 
 
     def KB_append(self, input):  ### inmemory documentstore please use KB construct
-        if self.retrieval_type == "dense":
-            if os.path.exists(input):
-                if os.path.isfile(input):
-                    data_collection = self.parse_document(input)
-                elif os.path.isdir(input):
-                    data_collection = self.batch_parse_document(input)
-                else:
-                    print("Please check your upload file and try again!")
+        if isinstance(input, str):
+            if os.path.isfile(input):
+                data_collection = self.parse_document(input)
+            elif os.path.isdir(input):
+                data_collection = self.batch_parse_document(input)
+            else:
+                print("Please check your upload file and try again!")
+        elif isinstance(input, List):
+            try:
+                data_collection = self.parse_html(input)
+            except:
+                print("The given link/str is unavailable. Please try another one!")
+        else:
+            print("The input format is invalid!")
 
+        if self.retrieval_type == "dense":
+            documents = []
+            for data, meta in data_collection:
+                if len(data) < 5:
+                    continue
+                metadata = {"source": meta}
+                new_doc = Document(page_content=data, metadata=metadata)
+                documents.append(new_doc)
+            assert documents != [], "The given file/files cannot be loaded."
+            vectordb = Chroma.from_documents(documents=documents, embedding=self.embeddings,
+                                             persist_directory=self.persist_dir)
+            vectordb.persist()
+            print("The local knowledge base has been successfully built!")
+            return Chroma(persist_directory=self.persist_dir, embedding_function=self.embeddings)
+        elif self.retrieval_type == "sparse":
+            if self.document_store == "Elasticsearch":
+                document_store = ElasticsearchDocumentStore(host="localhost", index=self.index_name,
+                                                            port=9200, search_fields=["content", "title"])
                 documents = []
                 for data, meta in data_collection:
+                    metadata = {"source": meta}
                     if len(data) < 5:

```

Comment:
same as above

Reply:
meta data here means the reference link/document_path of the data, we set a length filter of 5 to filter some noisy data whose length is to short. In the previous products the customer may provide lots of too short and no-means data to us

Reply:
define a global meaningful variable and refer to that

Reply:
> define a global meaningful variable and refer to that

will be fixed with the new vector database/retrieval/indexing change